src/DynamoCore/Utilities/NodeModelExtensions.cs

```diff
@@ -0,0 +1,144 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Autodesk.DesignScript.Interfaces;
+using Dynamo.DSEngine;
+using Dynamo.Models;
+using ProtoCore.Mirror;
+
+namespace Dynamo.Utilities
+{
+    public static class NodeModelExtensions
+    {
+        internal static void VisibleUpstreamNodes(this NodeModel node, List<NodeModel> gathered)
+        {
+            var upstream = node.InPorts.SelectMany(p => p.Connectors.Select(c => c.Start.Owner)).
+                ToList();
```

Comment:
No need for `ToList` here. `foreach` will simply iterate over the `IEnumerable` just as `ToList` would except we don't expand the list in memory.